rclpy/src/rclpy/_rclpy.c

```diff
@@ -6068,98 +6400,132 @@ static PyMethodDef rclpy_methods[] = {
-  if (NULL == RCLInvalidROSArgsError) {
-    Py_DECREF(m);
-    return NULL;
-  }
-  if (PyModule_AddObject(m, "RCLInvalidROSArgsError", RCLInvalidROSArgsError) != 0) {
-    Py_DECREF(m);
-    return NULL;
+  PyObject * exceptions_module = PyImport_ImportModuleLevel(
+    "_rclpy_pybind11", globals, NULL, NULL, 1);
+  if (NULL == exceptions_module) {
+    return -1;
   }
 
-  UnknownROSArgsError = PyErr_NewExceptionWithDoc(
-    "_rclpy.UnknownROSArgsError",
-    "Thrown when unknown ROS arguments are found.",
-    RCLError, NULL);
-  if (NULL == UnknownROSArgsError) {
-    Py_DECREF(m);
-    return NULL;
+  module_state->RCLError =
+    PyObject_GetAttrString(exceptions_module, "RCLError");
+  if (NULL == module_state->RCLError) {
+    Py_DECREF(exceptions_module);
+    return -1;
   }
-  if (PyModule_AddObject(m, "UnknownROSArgsError", UnknownROSArgsError) != 0) {
-    Py_DECREF(m);
-    return NULL;
+
+  module_state->RCLInvalidROSArgsError =
+    PyObject_GetAttrString(exceptions_module, "RCLInvalidROSArgsError");
+  if (NULL == module_state->RCLInvalidROSArgsError) {
+    Py_DECREF(module_state->RCLError);
+    Py_DECREF(exceptions_module);
+    return -1;
   }
 
-  NodeNameNonExistentError = PyErr_NewExceptionWithDoc(
-    "_rclpy.NodeNameNonExistentError",
-    "Thrown when a node name is not found.",
-    RCLError, NULL);
-  if (PyModule_AddObject(m, "NodeNameNonExistentError", NodeNameNonExistentError)) {
-    Py_DECREF(m);
-    return NULL;
+  module_state->UnknownROSArgsError =
+    PyObject_GetAttrString(exceptions_module, "UnknownROSArgsError");
+  if (NULL == module_state->UnknownROSArgsError) {
+    Py_DECREF(module_state->RCLInvalidROSArgsError);
+    Py_DECREF(module_state->RCLError);
+    Py_DECREF(exceptions_module);
+    return -1;
   }
 
-  UnsupportedEventTypeError = PyErr_NewExceptionWithDoc(
-    "_rclpy.UnsupportedEventTypeError",
-    "Thrown when registering a callback for an event type that is not supported.",
-    RCLError, NULL);
-  if (PyModule_AddObject(m, "UnsupportedEventTypeError", UnsupportedEventTypeError)) {
-    Py_DECREF(m);
-    return NULL;
+  module_state->NodeNameNonExistentError =
+    PyObject_GetAttrString(exceptions_module, "NodeNameNonExistentError");
+  if (NULL == module_state->NodeNameNonExistentError) {
+    Py_DECREF(module_state->UnknownROSArgsError);
+    Py_DECREF(module_state->RCLInvalidROSArgsError);
+    Py_DECREF(module_state->RCLError);
+    Py_DECREF(exceptions_module);
+    return -1;
   }
 
-  if (PyErr_Occurred()) {
-    Py_DECREF(m);
-    return NULL;
+  module_state->UnsupportedEventTypeError =
+    PyObject_GetAttrString(exceptions_module, "UnsupportedEventTypeError");
+  if (NULL == module_state->UnsupportedEventTypeError) {
+    Py_DECREF(module_state->NodeNameNonExistentError);
+    Py_DECREF(module_state->UnknownROSArgsError);
+    Py_DECREF(module_state->RCLInvalidROSArgsError);
+    Py_DECREF(module_state->RCLError);
+    Py_DECREF(exceptions_module);
```

Comment:
It's worth pointing out that these pointers aren't getting reset after the decrefs. Presumably Python will free the memory at its leisure, but I'm not sure what becomes of the module state, or if it sticks around for a subsequent attempt or is also freed.

Might be an opportunity to centralize this cleanup in a cleanup block at the end of the function, as is the pattern used throughout this file.

Reply:
IIUC an exception raised here becomes an exception raised [here](https://github.com/python/cpython/blob/6f8c8320e9eac9bc7a7f653b43506e75916ce8e8/Lib/importlib/util.py#L245), which means it will get thrown when something tries to access an attribute of the module.

The state itself exists on [the module PyObject](https://github.com/python/cpython/blob/6f8c8320e9eac9bc7a7f653b43506e75916ce8e8/Objects/moduleobject.c#L14) (the thing in `sys.modules["module.name"]`). [The module state is `free`'d when the module PyObject gets garbage collected](https://github.com/python/cpython/blob/6f8c8320e9eac9bc7a7f653b43506e75916ce8e8/Objects/moduleobject.c#L693). The module PyObject is going to stick around as long as there is a reference to it, which is usually forever unless someone calls [`importlib.reload()`](https://github.com/python/cpython/blob/6f8c8320e9eac9bc7a7f653b43506e75916ce8e8/Lib/importlib/__init__.py#L174). The state isn't reused, so no need to reset the pointers.